chore: codspeed benchmark

ematipico opened the pull request:
## Changes

This PR creates a new workflow to use continuous benchmarking using [CodSpeed](https://codspeed.io/withastro/astro)

Everything was set up months ago, but I couldn't follow up. 

I resurrected the code, tested it locally, and now opened it.

I created a new workflow because I don't want to disrupt too much the existing workflow we use for inline benchmarking via magic command.
Once we are in a place where CodSpeed is usable, I believe we can start using that one and ditch the ones that we use via command (only the workflow, of course).

## Testing

I tested the scripts locally, and they work. 

CI should pass. Then, upon merging the CI should trigger benchmarking and upload the results in the dashboard. The next, rebase PR, should show a bot message.

<!-- How was this change tested? -->
<!-- DON'T DELETE THIS SECTION! If no tests added, explain why. -->

## Docs

N/A

<!-- Could this affect a user’s behavior? We probably need to update docs! -->
<!-- If docs will be needed or you’re not sure, uncomment the next line: -->
<!-- /cc @withastro/maintainers-docs for feedback! -->

<!-- DON'T DELETE THIS SECTION! If no docs added, explain why.-->
<!-- https://github.com/withastro/docs -->

bluwy commented on benchmark/bench/codspeed.js:
```diff
@@ -0,0 +1,83 @@
+import { benchmarkCannon } from './server-stress.js';
+
+export async function run({ memory, render, stress }) {
+	const bench = process.env.CODE_SPEED ? withCodSpeed(new Bench()) : new Bench();
```
```suggestion
	const bench = process.env.CODSPEED ? withCodSpeed(new Bench()) : new Bench();
```

Should we match the env name with `codspeed`?

bluwy commented on benchmark/bench/server-stress.js:
```diff
@@ -27,9 +27,11 @@ export async function run(projectDir, outputFile) {
-	const previewProcess = execaCommand(`${astroBin} preview --port ${port}`, {
-		cwd: root,
-		stdio: 'inherit',
+	const previewProcess = exec([`${astroBin} preview --port ${port}`], {
```
Not sure how this one slipped away while refactoring to tinyexec. I think the correct usage here is something like this:
```suggestion
	const previewProcess = exec(astroBin, ['preview', '--port', port], {
```

bluwy commented on .github/workflows/continuous_benchmark.yml:
```diff
@@ -0,0 +1,50 @@
+  TURBO_TEAM: ${{ secrets.TURBO_TEAM }}
+  FORCE_COLOR: true
+  CODSPEED_TOKEN: ${{ secrets.CODSPEED_TOKEN }}
+  CODE_SPEED: true
```
This name would also need to be updated if we rename to `process.env.CODSPEED`

bluwy commented on benchmark/bench/codspeed.js:
```diff
@@ -0,0 +1,83 @@
+			});
+
+			console.info('Previewing...');
+			const previewProcess = await exec(astroBin, ['build'], {
```
The error you mentioned in the PR description and in CI is because this isn't previewing
```suggestion
			const previewProcess = await exec(astroBin, ['preview', '--port', port], {
```

bluwy commented on benchmark/bench/codspeed.js:
```diff
@@ -0,0 +1,83 @@
+			});
+
+			console.info('Previewing...');
+			const previewProcess = await exec(astroBin, ['build'], {
```
Same here
```suggestion
			const previewProcess = await exec(astroBin, ['preview', '--port', port], {
```

bluwy reviewed (commented):
I'd prefer if we can avoid duplicating the code for the benchmarks, but happy to try something out first.

Also a bit concerned with this running on each push to the pull request, maybe if can run only if there's a label. But also not a blocker

ematipico commented on benchmark/bench/cli-startup.js:
```diff
@@ -1,16 +1,15 @@
- * @param {URL} outputFile
  */
-export async function run(projectDir, outputFile) {
+export async function run(projectDir) {
```
The variable wasn't used

bluwy commented on benchmark/bench/codspeed.js:
```diff
@@ -0,0 +1,107 @@
+				const port = 4322;
+
+				console.info('Previewing...');
+				const previewProcess = exec(astroBin, ['preview', '--port', port], {
+					nodeOptions: {
+						cwd: render.root,
+						stdio: 'inherit',
+					},
+				});
```
From the CI logs, looks like this is still failing somehow. The console should log:

```
Preview server listening on http://localhost:4322
```

Maybe it's easier to skip this test for now, or to debug it you can pass `throwOnError: true` besides the `nodeOptions` option. 

I need to followup and improve the error logging for tinyexec because it doesn't throw for exit code 1 by default 😢

ematipico commented:
> Also a bit concerned with this running on each push to the pull request, maybe if can run only if there's a label. But also not a blocker

That would kill the purpose of this tool. Maybe we could evaluate a reduced workload for these benchmarks

ematipico commented:
This is weird, it works locally, but it fails in CI: https://github.com/withastro/astro/actions/runs/11683003640/job/32531309684?pr=12347

bluwy commented:
I tested and it also works locally. I can't tell what's going on here, there's no output or errors from the preview command at all. It's either the command not running at all, or it's hanging.

I'm also reading the tinybench code and it seems to be running the benchmark repeatedly instead of once, so locally it took very long to finish. I assume this is only temporary, but could a better way be using our own benchmark code and then reporting the results to codspeed ourselves manually?

ematipico commented:
> I'm also reading the tinybench code and it seems to be running the benchmark repeatedly instead of once, so locally it took very long to finish. I assume this is only temporary, but could a better way be using our own benchmark code and then reporting the results to codspeed ourselves manually?

This is actually expected by a benching library. Other libraries, in other ecosystems too, behave the same way. They run multiple times with the same function/task so they can create a sample out of it.

I'll try to use vitest at this point.

bluwy commented:
I guess that makes sense for our benchmarks like `cli-startup` and `memory`, but for `server-stress` and `render`, we already do the repeated runs ourselves so the result reported there should be stable to be used. Using tinybench for them is like running a benchmark system within another benchmark system 😄

What we also want is to measure are things like "average render time for MDX pages" or "requests/sec for Astro pages". If we wrap with tinybench, we're instead measuring "how long it takes to benchmark the 'average render time for MDX pages'" (for example).

I'm ok with testing `cli-startup` or `memory` to setup the initial codspeed workflows, since those don't have repeated runs. And then maybe later we can circle back on `server-stress` and `render` and refactor in a way that we can reuse the reported results, or rely on tinybench for the repeated runs? (And we can figure out the preview issue later).

ematipico commented:
I changed the approach now and removed the need for a preview server to test the rendering. I created an adapter that is very similar to our test adapter, we create a `Request` and call the `app.render`. 

This makes sense for rendering because it removes the noise of a Node.js HTTP server. With this approach, we use the minimal `NodeApp` and call `.render` straight away.

ematipico commented:
I am going to merge this for now. We will have a discussion internally about what to benchmark and how, so things might change
